File: tools/flightfinder/flightfinder/weather.py
"""Previsão/clima no destino nas datas do voo, via Open-Meteo (sem chave).

Datas a até ~14 dias usam previsão; datas distantes usam a média histórica do
mesmo período no ano anterior (rotulada como estimativa sazonal).
https://open-meteo.com
"""
from __future__ import annotations

import datetime as dt

from . import netutil

GEO = "https://geocoding-api.open-meteo.com/v1/search"
FORECAST = "https://api.open-meteo.com/v1/forecast"
ARCHIVE = "https://archive-api.open-meteo.com/v1/archive"
DAILY = "temperature_2m_max,temperature_2m_min,precipitation_sum"


def _geocode(city: str) -> tuple[float, float, str] | None:
    try:
        d = netutil.fetch_json(f"{GEO}?name={city.replace(' ', '+')}&count=1&language=pt", retries=1)
        res = (d.get("results") or [])
        if res:
            r = res[0]
            return float(r["latitude"]), float(r["longitude"]), r.get("name", city)
    except Exception:
        pass
    return None


def _avg(xs):
    xs = [x for x in xs if x is not None]
    return sum(xs) / len(xs) if xs else None


def summary(city: str, start_date: str, end_date: str | None = None) -> str | None:
    geo = _geocode(city)
    if not geo:
        return None
    lat, lon, name = geo
    end_date = end_date or start_date
    today = dt.date.today()
    try:
        start = dt.date.fromisoformat(start_date)
    except ValueError:
        return None
    ahead = (start - today).days

    if 0 <= ahead <= 14:
        url = (f"{FORECAST}?latitude={lat}&longitude={lon}&daily={DAILY}"
               f"&timezone=auto&start_date={start_date}&end_date={end_date}")
        label = "previsão"
    else:
        def prev(d):
            return (dt.date.fromisoformat(d).replace(year=dt.date.fromisoformat(d).year - 1)).isoformat()
        url = (f"{ARCHIVE}?latitude={lat}&longitude={lon}&daily={DAILY}"
               f"&timezone=auto&start_date={prev(start_date)}&end_date={prev(end_date)}")
        label = "média histórica (mesmo período do ano passado)"

    try:
        d = netutil.fetch_json(url, retries=1)
        daily = d.get("daily") or {}
        tmax = _avg(daily.get("temperature_2m_max") or [])
        tmin = _avg(daily.get("temperature_2m_min") or [])
        prec = daily.get("precipitation_sum") or []
        rain_days = sum(1 for p in prec if p and p > 1.0)
        if tmax is None:
            return None
        chuva = ("sem chuva relevante" if rain_days == 0
                 else f"{rain_days} dia(s) com chuva")
        return f"{name} ({label}): máx ~{tmax:.0f}°C, mín ~{tmin:.0f}°C, {chuva}"
    except Exception:
        return None

File: lib/sources/weather.ts
// Clima no destino nas datas do voo via Open-Meteo (sem chave). Datas a até ~14
// dias usam previsão; datas distantes usam média histórica do ano anterior.
import { fetchJson } from "@/lib/http";

const GEO = "https://geocoding-api.open-meteo.com/v1/search";
const FORECAST = "https://api.open-meteo.com/v1/forecast";
const ARCHIVE = "https://archive-api.open-meteo.com/v1/archive";
const DAILY = "temperature_2m_max,temperature_2m_min,precipitation_sum";

function avg(xs: (number | null | undefined)[]): number | null {
  const v = xs.filter((x): x is number => typeof x === "number");
  return v.length ? v.reduce((a, b) => a + b, 0) / v.length : null;
}

function prevYear(iso: string): string {
  const d = new Date(iso + "T00:00:00Z");
  d.setUTCFullYear(d.getUTCFullYear() - 1);
  return d.toISOString().slice(0, 10);
}

export async function weatherSummary(city: string, startDate: string, endDate?: string | null): Promise<string | null> {
  let geo: any;
  try {
    geo = await fetchJson<any>(`${GEO}?name=${encodeURIComponent(city)}&count=1&language=pt`, { timeout: 8000 });
  } catch {
    return null;
  }
  const r = (geo?.results || [])[0];
  if (!r) return null;
  const { latitude: lat, longitude: lon, name } = r;
  const end = endDate || startDate;

  const today = new Date();
  const start = new Date(startDate + "T00:00:00Z");
  const ahead = Math.round((start.getTime() - today.getTime()) / 86400000);

  let url: string;
  let label: string;
  if (ahead >= 0 && ahead <= 14) {
    url = `${FORECAST}?latitude=${lat}&longitude=${lon}&daily=${DAILY}&timezone=auto&start_date=${startDate}&end_date=${end}`;
    label = "previsão";
  } else {
    url = `${ARCHIVE}?latitude=${lat}&longitude=${lon}&daily=${DAILY}&timezone=auto&start_date=${prevYear(startDate)}&end_date=${prevYear(end)}`;
    label = "média histórica (mesmo período do ano passado)";
  }

  try {
    const d = await fetchJson<any>(url, { timeout: 9000 });
    const daily = d?.daily || {};
    const tmax = avg(daily.temperature_2m_max || []);
    const tmin = avg(daily.temperature_2m_min || []);
    const prec: number[] = daily.precipitation_sum || [];
    const rainDays = prec.filter((p) => p && p > 1).length;
    if (tmax == null) return null;
    const chuva = rainDays === 0 ? "sem chuva relevante" : `${rainDays} dia(s) com chuva`;
    return `${name} (${label}): máx ~${Math.round(tmax)}°C, mín ~${Math.round(tmin ?? 0)}°C, ${chuva}`;
  } catch {
    return null;
  }
}
